### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by Mradr** [gevent/gevent#1609]
Well, getting a strange behavior when running my Gevent application on a VM and hardware base machine running the newest 20.5.0 and python 3.7.7 in that with the application doing nothing at all - I seem to be getting HIGH CPU usage with one of the cores. For a 4 core system that would lead to 20-25% cpu usage. I don't really have any test code, but for the most part - I can comfirm that accross 3 different machines this seems to be a problem and I am only running one greenlet thread that sleeps 30 seconds in between anything.

Looks like this issue has been around for a little while - still testing:
python 3.7.7 - Windows 10 - base
20.5.0 -> high CPU usage
1.5.0 -> high CPU usage
1.4.0 -> normal 1-2%
1.3.7 -> normal 1-2%

Possible something change in between that I wasn't aware. I be more than welcome to do any extra testing that you would could take value from.

**Comment by jamadden** [comment]
I'm not aware of any specific issues, and don't see that in my own tests or production environments. (Recall that Windows is supported for development, not production.)

What does profiling show you? Something like [py-spy](https://pypi.org/project/py-spy/0.4.0.dev0/) that can sample native code too is ideal.

**Comment by Mradr** [comment]
If you don't mind me asking to hit two birds with one stone here then, what would be a better production product similar to Gevent? For the server, I am sure I could move it over to Linux no problem thus making full use, but for my client code, I just wanted something simple to use over the complexity of what was out there already.


Attach is the profile.svg

**Comment by jamadden** [comment]
> Attach is the profile.svg

Thanks! (I should have asked for the speedscope format 🤷 )

> TLDR - it mostly looks like sleep commands - witch is true for my code, but lots of time in the .loop.py found in gevent.

Its curious that none of the C code is showing up. I think you might have to specify an option explicitly to py-spy to get that? (I should have mentioned that too.)

Without the C code, but especially without seeing (a [SSCCE](http://sscce.org) version of) your code, I can't do much but guess., and all I can do there is suspect this example is running into some pathological corner case, such as huge numbers of callbacks/greenlets contesting over the same resource—there have been several changes from 1.4 to 1.5 that make gevent more fair and less likely to starve greenlets, or otherwise improve behaviour when the system is loaded. Some of that comes at a cost.

There were four pre-releases from 1.5a1 to 1.5a4 phasing in those changes. If you see this behaviour in one pre-release, but not the former, that would limit the scope of things to investigate.

> but for my client code, I just wanted something simple to use over the complexity of what was out there already.

For a simple client, you're most likely fine. 

Windows isn't recommended for production because I don't do that or know anyone who does, so there's lots of unknowns with regards to scalability, long-lived processes, etc. The event loop on Windows is also much less tolerant of programming errors than the loop on other processes, so it's easier to crash the interpreter accidentally.  Also, historically, a lot of gevent's functionality isn't possible on Windows; newer versions of Python, plus the use of libuv, are slowly improving that, but there are still things that just aren't available (e.g., prepare watchers, fork watchers and fork in general, off the top of my head).

**Comment by jamadden** [comment]
No updates since May so I'm closing. Feel free to reopen with more updates!
